src/rgw/rgw_lua_request.cc

```diff
@@ -244,7 +244,22 @@ struct ObjectMetaTable : public EmptyMetaTable {
   }
 };
 
-template<typename MapType=std::map<std::string, std::string>>
+typedef int MetaTableClosure(lua_State* L);
+
+template<typename MapType>
+int StringMapWriteableNewIndex(lua_State* L) {
+  const auto map = reinterpret_cast<MapType*>(lua_touserdata(L, lua_upvalueindex(1)));
+
+  ceph_assert(lua_isstring(L, -1));
+  ceph_assert(lua_isstring(L, -2));
+  const char* value = lua_tostring(L, -1);
+  const char* index = lua_tostring(L, -2);
```

Comment:
Using asserts for type checking should be avoided. Instead use Lua's exception mechanisms:

```suggestion
  const char* index = luaL_checkstring(L, 2);
```

Reply:
> Using asserts for type checking should be avoided. Instead use Lua's exception mechanisms:

I made the fix here, as well as other places in the code.
however, I had to keep the original indexing (-1, -2, etc.) for the code to work. the [manual](https://www.lua.org/manual/5.2/manual.html) says that this number should be a "function argument" - which is not clear to me.
anyway, passing the stack index there works.

Reply:
When the function is called from Lua, its arguments are pushed on the stack in order. So if a function is called with three arguments, you can refer to the first one either with `1` or `-3`. It's usually much clearer and better discipline to refer to function arguments by the absolute index (rather than a relative one, i.e. negative) but I'm not sure why it didn't work for you. The C++ gymnastics I haven't taken the time to follow. I'm used to reading regular C code Lua bindings.

Reply:
using the argument index works without any issue. changed the code to use parameter index instead of negative stack intexes.
my mistake was that the "index" parameter is the 2nd one, as the first parameter to the `__index()` function is the table name.